File: 5/main.ts
import { readFileSync } from 'fs'

let lines = readFileSync("input").toString().split("\r\n")

const dim = 1000

let grid: number[][] = []
for (let i = 0; i < dim; i++) {
  const y = []
  for (let j = 0; j < dim; j++) {
    y.push(0)
  }
  grid.push(y)
}

for (let line of lines) {
  const coords = line.split(" -> ")
  let [x1, y1] = coords[0].split(",")
  .map(coord => parseInt(coord))
  let [x2, y2] = coords[1].split(",")
  .map(coord => parseInt(coord))

  const x = [x1, x2]
  const y = [y1, y2]

  const insert = (x: number, y: number) => {
    grid[x][y]++
  }

  if (x[0] !== x[1] && y[0] !== y[1]) {
    for (let i = x[0], j = y[0]; x[0] < x[1] ? i <= x[1] : i >= x[1]; x[0] < x[1] ? i++ : i--,y[0] < y[1] ? j++ : j--) {
      insert(i, j)
    }
  } else if (x[0] !== x[1]) {
    x.sort((a, b) => a - b)
    y.sort((a, b) => a - b)
    for (let i = x[0]; i <= x[1]; i++) {
      insert(i, y[0])
    }
  } else if (y[0] !== y[1]) {
    x.sort((a, b) => a - b)
    y.sort((a, b) => a - b)
    for (let i = y[0]; i <= y[1]; i++) {
      insert(x[0], i)
    }
  }
}

const soln = grid.reduce((
  previousValue,
  currentValue,
) => previousValue + currentValue.filter(amount => amount >= 2).length, 0)
console.log("min 2:", soln)

File: 5/main.go
package main

import (
	"io"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
)

func map_(slice []int, mapFn func(int) int) []int {
	mapped := make([]int, len(slice))
	for i, elem := range slice {
		mapped[i] = mapFn(elem)
	}
	return mapped
}

func mapToInts(slice []string) ([]int, error) {
	mapped := make([]int, len(slice))
	for i, elem := range slice {
		conv, err := strconv.Atoi(elem)
		if err != nil {
			return nil, err
		}
		mapped[i] = conv
	}
	return mapped, nil
}

func reduce(slice []int, reducer func(prev int, cur int) int, initial int) int {
	prev := initial
	for _, elem := range slice {
		prev = reducer(prev, elem)
	}
	return prev
}

func filter(slice []int, filterFn func(int) bool) []int {
	var mapped []int
	for _, elem := range slice {
		if filterFn(elem) {
			mapped = append(mapped, elem)
		}
	}
	return mapped
}

func insert(grid [][]int, x int, y int) {
	grid[x][y]++
}

func main() {
	inputFile, err := os.Open("input")
	if err != nil {
		log.Fatalln(err)
	}

	fileContents, err := io.ReadAll(inputFile)
	if err != nil {
		log.Fatalln(err)
	}

	inputLines := strings.Split(string(fileContents), "\r\n")

	grid := make([][]int, 1000)

	for i := range grid {
		grid[i] = make([]int, 1000)
	}

	for _, line := range inputLines {
		coords := strings.Split(line, " -> ")

		c1, err := mapToInts(strings.Split(coords[0], ","))
		if err != nil {
			log.Fatalln(err)
		}
		c2, err := mapToInts(strings.Split(coords[1], ","))
		if err != nil {
			log.Fatalln(err)
		}

		x := []int{c1[0], c2[0]}
		y := []int{c1[1], c2[1]}

		sort.Ints()

		if line == "" {
			continue
		}
	}
}
